upd/Hamiltonian.py

```diff
@@ -31,12 +31,39 @@ def generate_two_body_integral(self, sym: int, basis: str, dense: bool):
         """
         pass
 
-    def to_dense(self, integral):
+    def to_sparse(self,Md):
         """
-        Generate a dense array from the sparse matrix
-        :param integral: sparse
-        :return: np.ndarray
+        Converts dense array of integrals to sparse array in dictionary of keys format
+        :param Md: dense array
+        :return Ms: sparse dictionary of keys
         """
+        # Find non-zero indices in dense array
+        indices = np.where(Md != 0)
+        # Building list object of indices
+        ind_ls = list(zip(*indices))
+        # Building dictionary of keys for non-zero dense array elements 
+        Ms = {key:value for key,value in zip(ind_ls, Md[indices])}
+        return Ms
+
+    def to_dense(self, Ms):
+        """
+        Converts sparse array of integrals in dictionary of keys format to dense numpy array
+        :param Ms: sparse dictionary of keys
+        :return: dense array
+        """
+        # Return 0 if sparse array is empty
+        if Ms == {}:
+            print("No non-zero integrals provided, dense array is 0.")
+            return 0
+        # Get shape of dense array from first element of sparse array
+        first = next(iter(Ms))
+        shape = list()
```

Comment:
Thanks @adsrichards this looks good! Here are minor suggestions:
1. I think it might be better to use list comprehension in here. `shape = [2*cm_len for x in first]`
Also, isn't it easier to find shape of an array by doing something like this: 
`shape = tuple(2*[2*self.one_mo.shape[0]])
`
2. It might be better to check if an array is `scipy.sparse.csr_matrix`, if so, we might just use `todense()` method available for them. This should work for one body integrals.

 Actually, it might be a good idea to store the shape of the sparse array as a key `shape` in dictionary and fill it during the populating. @msricher @gabrielasd what do you think?

Reply:
1. Thanks, the first expression for shape does look neater. For the second expression, I don't see how that would work as we could give a sparse matrix of the form {(1,2,3,4) : 1.0, (2,2,2,2) : 2.0}, while the size of the dense matrix could be 10x10x10x10 which we can't infer without the connectivity matrix size. 
2. Do we want to use scipy for sparse matrices? If so then I can put that in. My understanding was that they only work for 2D arrays which would give two different methods for the one and two-integral arrays.

Reply:
I think we'd decided to store only the (nonzero, underdetermined by whatever symmetry is specified by the user) elements, with explicit indexing and values. So there would be two arrays, one 2-dimensional integer array (nx2 or nx4 array, where n is the number of nonzero elements and there are either 2 (or 4) indices for one-electron (or 2-electron) integrals? The second array would have one float (or I guess it could be complex) for each entry in the first array. 

That seems the simplest data structure, though it is also sensible to make a linked list (because the number of elements in the array is not certain from the beginning) and then only convert to an array if/when it is needed for output.

Reply:
> 

Can we store the one body integral as a Scipy.sparase.csr_matrix? It is always 2d array and there shouldn't be any issues with converting to sparse format (so we can handle with it as a regular numpy array in the code). 
As for the data structure for two body integrals we store them as a dictionary in which keys are indices of non zero elements and values are the values of the corresponding elements in the array. In this way we don't need to have 2 arrays for the one object 


@adsrichards I definitely agree that we need to know the dimensionality and we can get it from the connectivity matrix.  I'm not sure what data type do you wanna store at self.cm_len in case it's a shape of connectivity matrix than we can just do: `shape = tuple(4*[2*self.cnct_matrxi.shape[0]])` which returns us tuple of 4 indices that are 2x larger than shape of connectivity matrix. Will it work better @adsrichards ?

Reply:
For 1-electron integrals works perfect to use scipy.sparse. For two-electron integrals the dictionary is quite fine; it automatically handles the caching issues too.My only questions is whether it is confusing to use scipy.sparse for one-electron integrals but a dictionary for 2-electron integrals; one could argue for using dictionaries for both I guess. But one can make good arguments for both....